### ISSUE TITLE
CKAN 2.2.1 to 2.5.1 upgrade: "column system_info.state does not exist" - DB migrations not working?

### ISSUE THREAD

**Issue by RJacksonm1** [ckan/ckan#2876]
Hi there,

I'm upgrading a CKAN 2.2.1 installation to 2.5.1.  I'm currently stuck with the following when I try to access the site:

```

[Thu Feb 11 17:04:02.140942 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823] mod_wsgi (pid=3042): Target WSGI script '/etc/ckan/default/apache.wsgi' cannot be loaded as Python module.
[Thu Feb 11 17:04:02.141049 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823] mod_wsgi (pid=3042): Exception occurred processing WSGI script '/etc/ckan/default/apache.wsgi'.
[Thu Feb 11 17:04:02.141109 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823] Traceback (most recent call last):
[Thu Feb 11 17:04:02.141164 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/etc/ckan/default/apache.wsgi", line 9, in <module>
[Thu Feb 11 17:04:02.141246 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     application = loadapp('config:%s' % config_filepath)
[Thu Feb 11 17:04:02.141304 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/local/lib/python2.7/dist-packages/paste/deploy/loadwsgi.py", line 247, in loadapp
[Thu Feb 11 17:04:02.141367 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     return loadobj(APP, uri, name=name, **kw)
[Thu Feb 11 17:04:02.141409 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/local/lib/python2.7/dist-packages/paste/deploy/loadwsgi.py", line 272, in loadobj
[Thu Feb 11 17:04:02.141460 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     return context.create()
[Thu Feb 11 17:04:02.522553 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/local/lib/python2.7/dist-packages/paste/deploy/loadwsgi.py", line 710, in create
[Thu Feb 11 17:04:02.522710 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     return self.object_type.invoke(self)
[Thu Feb 11 17:04:02.522794 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/local/lib/python2.7/dist-packages/paste/deploy/loadwsgi.py", line 146, in invoke
[Thu Feb 11 17:04:02.522915 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     return fix_call(context.object, context.global_conf, **context.local_conf)
[Thu Feb 11 17:04:02.523000 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/local/lib/python2.7/dist-packages/paste/deploy/util.py", line 56, in fix_call
[Thu Feb 11 17:04:02.523092 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     val = callable(*args, **kw)
[Thu Feb 11 17:04:02.523170 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/opt/ckan/ckan2.5.1/lib/default/src/ckan/ckan/config/middleware.py", line 63, in make_app
[Thu Feb 11 17:04:02.523264 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     app_globals.reset()
[Thu Feb 11 17:04:02.523344 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/opt/ckan/ckan2.5.1/lib/default/src/ckan/ckan/lib/app_globals.py", line 167, in reset
[Thu Feb 11 17:04:02.523435 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     get_config_value(key)
[Thu Feb 11 17:04:02.523508 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/opt/ckan/ckan2.5.1/lib/default/src/ckan/ckan/lib/app_globals.py", line 135, in get_config_value
[Thu Feb 11 17:04:02.523591 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     value = model.get_system_info(key)
[Thu Feb 11 17:04:02.523662 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/opt/ckan/ckan2.5.1/lib/default/src/ckan/ckan/model/system_info.py", line 54, in get_system_info
[Thu Feb 11 17:04:02.523749 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     obj = meta.Session.query(SystemInfo).filter_by(key=key).first()
[Thu Feb 11 17:04:02.523827 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2334, in first
[Thu Feb 11 17:04:02.523921 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     ret = list(self[0:1])
[Thu Feb 11 17:04:02.523992 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2201, in __getitem__
[Thu Feb 11 17:04:02.524074 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     return list(res)
[Thu Feb 11 17:04:02.524144 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2405, in __iter__
[Thu Feb 11 17:04:02.524227 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     return self._execute_and_instances(context)
[Thu Feb 11 17:04:02.524297 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2420, in _execute_and_instances
[Thu Feb 11 17:04:02.524385 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     result = conn.execute(querycontext.statement, self._params)
[Thu Feb 11 17:04:02.524456 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 727, in execute
[Thu Feb 11 17:04:02.524568 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     return meth(self, multiparams, params)
[Thu Feb 11 17:04:02.524639 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 322, in _execute_on_connection
[Thu Feb 11 17:04:02.524726 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     return connection._execute_clauseelement(self, multiparams, params)
[Thu Feb 11 17:04:02.524797 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 824, in _execute_clauseelement
[Thu Feb 11 17:04:02.524884 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     compiled_sql, distilled_params
[Thu Feb 11 17:04:02.524953 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 954, in _execute_context
[Thu Feb 11 17:04:02.525038 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     context)
[Thu Feb 11 17:04:02.525115 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1116, in _handle_dbapi_exception
[Thu Feb 11 17:04:02.525202 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     exc_info
[Thu Feb 11 17:04:02.525271 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 189, in raise_from_cause
[Thu Feb 11 17:04:02.525388 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     reraise(type(exception), exception, tb=exc_tb)
[Thu Feb 11 17:04:02.525458 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 947, in _execute_context
[Thu Feb 11 17:04:02.525543 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     context)
[Thu Feb 11 17:04:02.525618 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]   File "/usr/lib/ckan/default/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 435, in do_execute
[Thu Feb 11 17:04:02.525702 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]     cursor.execute(statement, parameters)
[Thu Feb 11 17:04:02.525875 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823] ProgrammingError: (ProgrammingError) column system_info.state does not exist
[Thu Feb 11 17:04:02.525941 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823] LINE 1: ...info_key, system_info.value AS system_info_value, system_inf...
[Thu Feb 11 17:04:02.525997 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]                                                              ^
[Thu Feb 11 17:04:02.526051 2016] [wsgi:error] [pid 3042:tid 140452511594240] [remote 176.35.4.241:6823]  'SELECT system_info.id AS system_info_id, system_info.key AS system_info_key, system_info.value AS system_info_value, system_info.state AS system_info_state, system_info.revision_id AS system_info_revision_id \\nFROM system_info \\nWHERE system_info.key = %(key_1)s \\n LIMIT %(param_1)s' {'key_1': 'ckan.site_description', 'param_1': 1}

```

I'm wondering if this means the database migrations are not running for this upgrade. I have to switch my Solr schema back to 2.0 in order to run `db upgrade`, but when I do so it executes without any output.  I'm not sure if this is expected behaviour or not.

```
(default) root@duck:/etc/solr/conf# paster --plugin=ckan db upgrade --config=/etc/ckan/default/development.ini
(default) root@duck:/etc/solr/conf#
```

Once I run that, I switch Solr back to using the 2.3 schema (otherwise the site throws wrong-schema-version errors), and then I get stuck with the same "column does not exist" error as above.

Any pointers how I might investigate this further?

**Comment by rossjones** [comment]
I noticed you're using the development.ini for the db upgrade, are you also using that from apache?
Do production.ini and development.ini point to the same database?

**Comment by RJacksonm1** [comment]
We're using `production.ini` for Apache.  Both `development.ini` and `production.ini` point to the same database.  I tried the DB upgrade with both files, but neither appear to have resolved the problem.

In the background, on a cloned CKAN installation, I'm also trying the upgrade in smaller steps.  I've upgraded from `2.2.1` to `2.3.0`, though there are a few niggles for me to work out with that.  I'll keep playing around and update here if I find anything that might explain the problem with the `2.2.1` to `2.5.1` jump.

**Comment by RJacksonm1** [comment]
I have been successful in upgrading in smaller steps, from from `2.2.1` to `2.3.0` to `2.4.2` to `2.5.1`.  The only issues I'm encountering now appear to be related to our custom CKAN theme.

As I can work around this particular problem by upgrading in smaller steps, shall we consider this ticket closed?

**Comment by pduchesne** [comment]
Just in case this gets reopened in the future, here's some extra info, as I've run into this also while migrating from 2.2 to 2.6 : 

the bug is similar to #2029, but for system_info table, rather than user.
The fix for #2029 does not apply though, because this variant of the bug happens before the fix, when ckan/lib/cli.py calls `load_environment`. This happens since 048632d73588ed462a1133e796681056fe89ef37, where a call to `app_globals.reset()` has been introduced.

That said, I eventually also resorted to migrating the DB by smaller steps.
